frontend/src/app/models/create-model/create-model.component.ts

```diff
@@ -55,28 +62,50 @@ export class CreateModelComponent implements OnInit {
 
   ngOnInit(): void {
     this.toolService.get_tools().subscribe();
-    this.route.params.subscribe((params) => {
-      this.modelService.initAll(params.project).subscribe((models) => {
-        this.form.controls.name.addValidators(
-          this.slugValidator(models.map((value) => value.slug))
-        );
-      });
+    this.modelService._models.pipe(filter(Boolean)).subscribe((models) => {
+      this.form.controls.name.addValidators(
+        this.slugValidator(models.map((model) => model.slug))
+      );
     });
   }
 
   onSubmit(): void {
-    if (this.form.valid && this.projectService.project?.slug) {
+    if (this.form.valid && this.projectService.project!.slug) {
       let new_model = this.form.value as NewModel;
-      this.modelService
-        .createNewModel(this.projectService.project?.slug, new_model)
-        .subscribe((result) => {
-          this.router.navigate([
-            '/choose-source',
-            this.projectService.project?.slug,
-            result.slug,
-          ]);
+
+      const model_creation_subject = connectable<Model>(
+        this.modelService.createNewModel(
+          this.projectService.project!.slug,
+          new_model
+        ),
+        {
+          connector: () => new Subject(),
+          resetOnDisconnect: false,
+        }
+      );
+
+      model_creation_subject.subscribe((model) => {
+        this.router.navigate([
+          'project',
+          this.projectService.project!.slug,
+          'model',
+          model.slug,
+          'choose-source',
+        ]);
+      });
+
+      model_creation_subject
+        .pipe(
+          switchMap(() =>
+            this.modelService.list(this.projectService.project!.slug)
+          )
+        )
+        .pipe(single())
+        .subscribe({
+          next: this.modelService._models.next.bind(this.modelService._models),
```

Comment:
I list all models, then the single operator checks that the http call returns only one response (and not model, the response is an array of models). This operator is actually useless, because this http request always receive one response. Finally, I pass the next function of the _models BehaviorSubject, but since I use it outside the service, I need to specify the “this argument” of the function with .bind.
I think I can simplify the syntax there.

Reply:
Thanks for the explanation! 👍